### PR TITLE
Implement "Exact" HTTPPathMatch

### PR DESCRIPTION
**Pull request by nak3**
This patch adds "Exact" HTTPPathMatch.

Signed-off-by: Kenjiro Nakayama <nakayamakenjiro@gmail.com>

/cc @stevesloka @youngnick

### CODE REVIEW

File: internal/dag/gatewayapi_processor.go [hunk 1]
```diff
@@ -120,13 +120,15 @@ func (p *GatewayAPIProcessor) computeHTTPRoute(route *gatewayapi_v1alpha1.HTTPRo
 
 	for _, rule := range route.Spec.Rules {
 
-		var pathPrefixes []string
+		var pathMatchConditaions []MatchCondition
```
Comment (stevesloka) [inline internal/dag/gatewayapi_processor.go:123]: nit: s/pathMatchConditaions/pathMatchConditions/g

File: internal/dag/gatewayapi_processor.go [hunk 2]
```diff
@@ -120,13 +120,15 @@ func (p *GatewayAPIProcessor) computeHTTPRoute(route *gatewayapi_v1alpha1.HTTPRo
...
 
 		for _, match := range rule.Matches {
 			switch match.Path.Type {
 			case gatewayapi_v1alpha1.PathMatchPrefix:
-				pathPrefixes = append(pathPrefixes, stringOrDefault(match.Path.Value, "/"))
+				pathMatchConditaions = append(pathMatchConditaions, &PrefixMatchCondition{Prefix: stringOrDefault(match.Path.Value, "/")})
+			case gatewayapi_v1alpha1.PathMatchExact:
+				pathMatchConditaions = append(pathMatchConditaions, &ExactMatchCondition{Path: stringOrDefault(match.Path.Value, "/")})
 			default:
 				p.Error("NOT IMPLEMENTED: Only PathMatchPrefix is currently implemented.")
```
Comment (stevesloka) [inline internal/dag/gatewayapi_processor.go:133]: Can we update this error message?

**Review by stevesloka — CHANGES_REQUESTED** [review]

**Review by sunjayBhatia — COMMENTED** [review]
See also: https://github.com/projectcontour/contour/pull/3471 which does the same for ingress

there are some pieces in there that overlap with this change and are needed here, namely the sorting and other test changes

We should probably also add some integration testing to make sure this works end to end

**Comment by youngnick** [comment]
@nak3, do you think you could check over #3471 and see if anything needs updating here?

**Comment by nak3** [comment]
Umm... It seems sorting are moved to https://github.com/projectcontour/contour/pull/3504. The sort dag is already called by gateway APIs builder or this PR needs to be changed for the sort dag? @sunjayBhatia

**Comment by sunjayBhatia** [comment]
> Umm... It seems sorting are moved to #3504. The sort dag is already called by gateway APIs builder or this PR needs to be changed for the sort dag? @sunjayBhatia

Ah yeah forgot to update here when I pulled the sorting PR out as its own thing. You shouldn't need to change anything to accommodate sorting route matches in this PR, #3504 should take care of it and it was already merged.

**Comment by nak3** [comment]
Thank you!
Then I think this PR is ready now. cc @stevesloka @youngnick 
For e2e test, I think I should add it to contour-operator.

**Review by stevesloka — APPROVED** [review]

**Comment by nak3** [comment]
Rebased to solve confliction.

**Comment by sunjayBhatia** [comment]
> Rebased to solve confliction.
Looks like a CI failure to check out

**Comment by nak3** [comment]
Sorry I missed it. Update it now. Thank you!

**Review by sunjayBhatia — APPROVED** [review]

**Review by skriss — COMMENTED** [review]
Thanks for this @nak3, the code LGTM. It would be great to also add some test cases for exact matching to `_integration/testsuite/gatewayapi/001-patch-condition-match.yaml`.

**Comment by nak3** [comment]
Sure, added integration test. Thank you!

**Pull request merged**